From f5074d96d7f8930f9a17ff3299682e80a54cdacd Mon Sep 17 00:00:00 2001
From: Aaron Krajeski <aaronhk@chromium.org>
Date: Tue, 17 Dec 2024 10:24:44 -0800
Subject: [PATCH] Warn for unsupported gradient color spaces

These four spaces are in the spec, but are unsupported by Skia for
interpolation. We either need to implement them in Skia or inject a
shader. Before this CL the code reaches a NOTREACHED in
Blink::Gradient::ResolveSkInterpolation() and crashes (see linked bug):

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/graphics/gradient.cc;l=243?q=gradient.cc&sq=package:chromium&ct=os

For the time being, warn the dev and fail as if it's a parsing failure.

Bug: 382086298
Change-Id: Ie47a1caee682e021fdf5df331c50a6e49fcfa050
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6085127
Reviewed-by: Nico Weber <thakis@chromium.org>
Reviewed-by: Peter Boström <pbos@chromium.org>
Commit-Queue: Aaron Krajeski <aaronhk@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1397419}
---

--- a/third_party/blink/renderer/core/css/properties/css_parsing_utils.cc
+++ b/third_party/blink/renderer/core/css/properties/css_parsing_utils.cc
@@ -2735,6 +2735,18 @@
              : nullptr;
 }
 
+static void MaybeLogUnsupportedGradientInterpolationSpaceWarning(
+    const CSSParserContext& context,
+    Color::ColorSpace color_space) {
+  if (const auto* document = context.GetDocument()) {
+    document->AddConsoleMessage(MakeGarbageCollected<ConsoleMessage>(
+        mojom::blink::ConsoleMessageSource::kRecommendation,
+        mojom::blink::ConsoleMessageLevel::kWarning,
+        Color::ColorSpaceToString(color_space) +
+            " is not yet supported as a gradient interpolation space."));
+  }
+}
+
 static CSSValue* ConsumeRadialGradient(CSSParserTokenStream& stream,
                                        const CSSParserContext& context,
                                        cssvalue::CSSGradientRepeat repeating) {
@@ -2845,6 +2857,11 @@
           vertical_size, repeating, cssvalue::kCSSRadialGradient);
 
   if (has_color_space) {
+    if (Color::IsUndefinedColorSpaceForGradientInterpolation(color_space)) {
+      MaybeLogUnsupportedGradientInterpolationSpaceWarning(context,
+                                                           color_space);
+      return nullptr;
+    }
     result->SetColorInterpolationSpace(color_space, hue_interpolation_method);
     context.Count(WebFeature::kCSSColorGradientColorSpace);
   }
@@ -2913,6 +2930,11 @@
           end_x, end_y, nullptr, nullptr, angle, repeating, gradient_type);
 
   if (has_color_space) {
+    if (Color::IsUndefinedColorSpaceForGradientInterpolation(color_space)) {
+      MaybeLogUnsupportedGradientInterpolationSpaceWarning(context,
+                                                           color_space);
+      return nullptr;
+    }
     result->SetColorInterpolationSpace(color_space, hue_interpolation_method);
     context.Count(WebFeature::kCSSColorGradientColorSpace);
   }
@@ -2965,6 +2987,11 @@
       center_x, center_y, from_angle, repeating);
 
   if (has_color_space) {
+    if (Color::IsUndefinedColorSpaceForGradientInterpolation(color_space)) {
+      MaybeLogUnsupportedGradientInterpolationSpaceWarning(context,
+                                                           color_space);
+      return nullptr;
+    }
     result->SetColorInterpolationSpace(color_space, hue_interpolation_method);
     context.Count(WebFeature::kCSSColorGradientColorSpace);
   }
--- a/third_party/blink/renderer/platform/graphics/color.h
+++ b/third_party/blink/renderer/platform/graphics/color.h
@@ -117,6 +117,16 @@
     kNone,
   };
 
+  // TODO(crbug.com/383270111): As of 2024-12-12 Skia does not yet support these
+  // color spaces for gradient interpolation.
+  static bool IsUndefinedColorSpaceForGradientInterpolation(
+      ColorSpace color_space) {
+    return color_space == ColorSpace::kDisplayP3 ||
+           color_space == ColorSpace::kA98RGB ||
+           color_space == ColorSpace::kProPhotoRGB ||
+           color_space == ColorSpace::kRec2020;
+  }
+
   // For testing purposes and for serializer.
   static WTF::String ColorSpaceToString(Color::ColorSpace color_space);
 
--- /dev/null
+++ b/third_party/blink/web_tests/fast/css/gradient-unsupported-color-spaces-expected.txt
@@ -0,0 +1,13 @@
+CONSOLE WARNING: display-p3 is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: a98-rgb is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: rec2020 is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: prophoto-rgb is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: display-p3 is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: a98-rgb is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: rec2020 is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: prophoto-rgb is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: display-p3 is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: a98-rgb is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: rec2020 is not yet supported as a gradient interpolation space.
+CONSOLE WARNING: prophoto-rgb is not yet supported as a gradient interpolation space.
+
--- /dev/null
+++ b/third_party/blink/web_tests/fast/css/gradient-unsupported-color-spaces.html
@@ -0,0 +1,19 @@
+<!DOCTYPE html
+><body></body>
+<script>
+  if (window.testRunner)
+   testRunner.dumpAsText();
+
+  const gradientTypes = ["linear", "conic", "radial"];
+  // TODO(crbug.com/383270111): These color spaces are valid by css, but not yet
+  // supported as interpolation spaces for gradients by Skia.
+  const unsupporedColorSpaces = ["display-p3", "a98-rgb", "rec2020", "prophoto-rgb"];
+
+  for (gradientType of gradientTypes) {
+    for (colorSpace of unsupporedColorSpaces) {
+      gradient = `${gradientType}-gradient(in ${colorSpace}, red, blue)`;
+      document.body.style.background = gradient;
+    }
+  }
+</script>
+</html>
\ No newline at end of file
